File: lib/data/repositories/distance_matrix_repository.dart
import 'package:drift/drift.dart';

import '../../domain/models/distance_matrix_entry.dart';
import '../../infra/db/app_database.dart';

class DistanceMatrixRepository {
  final AppDatabase _db;
  DistanceMatrixRepository(this._db);

  Future<void> upsertMany(List<DistanceMatrixEntry> entries) async {
    await _db.batch((batch) {
      for (final e in entries) {
        batch.insert(
          _db.distanceMatrixTable,
          DistanceMatrixTableCompanion.insert(
            fromId: e.fromId,
            toId: e.toId,
            distanceMeters: e.distanceMeters,
            durationSeconds: e.durationSeconds,
            computedAt: e.computedAt.millisecondsSinceEpoch,
          ),
          mode: InsertMode.insertOrReplace,
        );
      }
    });
  }

  Future<int?> distanceMeters({required int from, required int to}) async {
    final row = await (_db.select(_db.distanceMatrixTable)
          ..where((t) => t.fromId.equals(from) & t.toId.equals(to)))
        .getSingleOrNull();
    return row?.distanceMeters;
  }

  Future<int?> durationSeconds({required int from, required int to}) async {
    final row = await (_db.select(_db.distanceMatrixTable)
          ..where((t) => t.fromId.equals(from) & t.toId.equals(to)))
        .getSingleOrNull();
    return row?.durationSeconds;
  }

  Future<List<DistanceMatrixEntry>> distancesFromPivot({
    required int pivotId,
    required int maxDistanceMeters,
  }) async {
    final rows = await (_db.select(_db.distanceMatrixTable)
          ..where((t) =>
              t.fromId.equals(pivotId) &
              t.distanceMeters.isSmallerOrEqualValue(maxDistanceMeters) &
              t.toId.equals(pivotId).not() &
              t.toId.equals(DistanceMatrixEntry.baseId).not()))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<void> deleteForClient(int clientId) async {
    await (_db.delete(_db.distanceMatrixTable)
          ..where((t) =>
              t.fromId.equals(clientId) | t.toId.equals(clientId)))
        .go();
  }

  Future<void> deleteAll() async {
    await _db.delete(_db.distanceMatrixTable).go();
  }

  Future<int> countForNode(int id) async {
    final result = await _db.customSelect(
      'SELECT COUNT(*) AS c FROM distance_matrix '
      'WHERE from_id = ? OR to_id = ?',
      variables: [Variable.withInt(id), Variable.withInt(id)],
      readsFrom: {_db.distanceMatrixTable},
    ).getSingle();
    return result.data['c'] as int;
  }

  DistanceMatrixEntry _toDomain(DistanceMatrixRow row) => DistanceMatrixEntry(
        fromId: row.fromId,
        toId: row.toId,
        distanceMeters: row.distanceMeters,
        durationSeconds: row.durationSeconds,
        computedAt:
            DateTime.fromMillisecondsSinceEpoch(row.computedAt),
      );
}

File: src/data/repositories/distance-matrix-repository.ts
import { and, eq, lt } from 'drizzle-orm';
import { distanceMatrix } from '@/infra/db/schema';
import type { Db } from '@/infra/db/client';
import { DistanceMatrixEntry } from '@/domain/models/distance-matrix-entry';

interface DistanceMatrixRow {
  fromId: string;
  toId: string;
  distanceKm: number;
  durationMinutes: number;
  fetchedAt: string;
}

function fromRow(r: DistanceMatrixRow): DistanceMatrixEntry {
  return DistanceMatrixEntry.parse(r);
}

export class DistanceMatrixRepository {
  constructor(private readonly db: Db) {}

  async byPair(fromId: string, toId: string): Promise<DistanceMatrixEntry | null> {
    const rows = await this.db
      .select()
      .from(distanceMatrix)
      .where(and(eq(distanceMatrix.fromId, fromId), eq(distanceMatrix.toId, toId)));
    return rows[0] ? fromRow(rows[0] as DistanceMatrixRow) : null;
  }

  async upsert(entry: DistanceMatrixEntry): Promise<void> {
    await this.db
      .insert(distanceMatrix)
      .values(entry)
      .onConflictDoUpdate({
        target: [distanceMatrix.fromId, distanceMatrix.toId],
        set: entry,
      });
  }

  async upsertMany(entries: DistanceMatrixEntry[]): Promise<void> {
    for (const e of entries) {
      await this.upsert(e);
    }
  }

  async deleteOlderThan(isoDate: string): Promise<void> {
    await this.db.delete(distanceMatrix).where(lt(distanceMatrix.fetchedAt, isoDate));
  }

  async listAll(): Promise<DistanceMatrixEntry[]> {
    const rows = await this.db.select().from(distanceMatrix);
    return rows.map((r) => fromRow(r as DistanceMatrixRow));
  }
}
